extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,288 @@
+import com.google.common.primitives.Ints;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import net.jpountz.lz4.LZ4Exception;
+import org.apache.commons.codec.digest.DigestUtils;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+import redis.clients.jedis.exceptions.JedisException;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * @author estliberitas
+ */
+public class RedisCache implements Cache
+{
+  private static final Logger log = new Logger(RedisCache.class);
+  private final JedisPool pool;
+  private final String prefix;
+  private final long expiration;
+  private final RedisTranscoder transcoder;
+  private final AtomicLong hitCount = new AtomicLong(0);
+  private final AtomicLong missCount = new AtomicLong(0);
+  private final AtomicLong timeoutCount = new AtomicLong(0);
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    JedisPoolConfig poolConfig = config.getPoolConfig();
+    JedisPool pool = new JedisPool(poolConfig, config.getHost(), config.getPort(), config.getTimeout());
+    return new RedisCache(pool, config);
+  }
+
+  public RedisCache(JedisPool pool, RedisCacheConfig config)
+  {
+    this.pool = pool;
+    prefix = config.getPrefix();
+    expiration = config.getExpiration();
+    transcoder = new RedisTranscoder();
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    final byte[] hash = computeKeyHash(prefix, key);
+    byte[] result = null;
+
+    try (final Jedis client = pool.getResource()) {
+      final byte[] serialized = client.get(hash);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] hash = computeKeyHash(prefix, key);
+    final byte[] serialized;
+
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (final Jedis client = pool.getResource()) {
+      client.psetex(hash, expiration, serialized);
```

Comment:
Also, if we were to not set any expiration and configure redis using `maxmemory-policy allkeys-lru`, then we don't need to set expiration time, and can save ourselves a few bytes of memory.